src/coreclr/jit/valuenum.cpp

```diff
@@ -8138,12 +8138,22 @@ void Compiler::fgValueNumberTree(GenTree* tree)
             if (tree->gtFlags & GTF_IND_INVARIANT)
             {
                 assert(!isVolatile); // We don't expect both volatile and invariant
-                tree->gtVNPair =
-                    ValueNumPair(vnStore->VNForMapSelect(VNK_Liberal, TYP_REF, ValueNumStore::VNForROH(),
-                                                         addrNvnp.GetLiberal()),
-                                 vnStore->VNForMapSelect(VNK_Conservative, TYP_REF, ValueNumStore::VNForROH(),
-                                                         addrNvnp.GetConservative()));
-                tree->gtVNPair = vnStore->VNPWithExc(tree->gtVNPair, addrXvnp);
+
+                // Is it a string literal? (it's always non-null)
+                if (addr->IsCnsIntOrI() && addr->IsIconHandle(GTF_ICON_STR_HDL))
+                {
+                    tree->gtVNPair = vnStore->VNPairForFunc(tree->TypeGet(), VNF_StrCns, addrNvnp);
+                    tree->gtVNPair = vnStore->VNPWithExc(tree->gtVNPair, addrXvnp);
```

Comment:
I think if it is an `Indir(const string ptr)` it can't have any exceptions, so we can `assert(addrXvnp == vnStore->VNForEmptyExcSet())` and skip `tree->gtVNPair = vnStore->VNPWithExc(tree->gtVNPair, addrXvnp);`.

Reply:
Make sense, thanks!

Reply:
sorry, I made a mistake, it should be `assert(addrXvnp == vnStore->VNPForEmptyExcSet())` , not `VNForEmptyExcSet`